src/Rails.elm

```diff
@@ -1,300 +1,395 @@
+            , expect = Rails.expectEmptyBody HandleDeletedHatResponse
             , timeout = Nothing
-            , withCredentials = False
+            , tracker = Nothing
             }
 
 **NOTE:** Rails typically expects an `X-CSRF-Token` header for requests other
 than `GET`, which this does not include. One way to have this header included
-automatically is to add [`csrf-xhr`](https://www.npmjs.com/package/csrf-xhr)
-to the page, before your `Elm` program gets initialized.
+automatically is to add [`csrf-xhr`](https://www.npmjs.com/package/csrf-xhr) to
+the page, before your `Elm` program gets initialized.
 
 -}
 request :
     { method : String
     , headers : List Header
     , url : String
     , body : Body
-    , expect : Expect a
+    , expect : Expect msg
     , timeout : Maybe Float
-    , withCredentials : Bool
+    , tracker : Maybe String
     }
-    -> Request a
-request options =
+    -> Cmd msg
+request { method, headers, url, body, expect, timeout, tracker } =
     let
-        headers =
-            List.concat
-                [ defaultRequestHeaders
-                , options.headers
-                ]
+        accept =
+            case expect of
+                ExpectJson _ ->
+                    -- q indicates our preference, from 0 to 1. We almost never
+                    -- want a fallback, so we indicate that.
+                    Http.header "Accept" "application/json, text/javascript, */*; q=0.01"
+
+                ExpectString _ ->
+                    Http.header "Accept" "*/*"
+
+        requestedWith =
+            Http.header "X-Requested-With" "XMLHttpRequest"
+
+        unwrappedExpect =
+            case expect of
+                ExpectJson inner ->
+                    inner
+
+                ExpectString inner ->
+                    inner
     in
-    Http.request { options | headers = headers }
+    Http.request
+        { method = method
+        , headers = accept :: requestedWith :: headers
+        , url = url
+        , body = body
+        , expect = unwrappedExpect
+        , timeout = timeout
+        , tracker = tracker
+        }
 
 
-defaultRequestHeaders : List Header
-defaultRequestHeaders =
-    [ Http.header "Accept" "application/json, text/javascript, */*; q=0.01"
-    , Http.header "X-Requested-With" "XMLHttpRequest"
-    ]
 
+-- EXPECTATIONS
 
-{-| Decode Rails-specific error information from a [`BadStatus`](http://package.elm-lang.org/packages/elm-lang/http/latest/Http#Error)
-response. (That is, a response whose status code is outside the 200 range.)
 
-This is intended to be used with [`Http.send`](http://package.elm-lang.org/packages/elm-lang/http/1.0.0/Http#send)
-like so:
+{-| Expect that the response body will look a certain way. Similar to
+[`Http.Expect`](https://package.elm-lang.org/packages/elm/http/2.0.0/Http#Expect),
+but wrapped here so we know what headers to set in requests.
+-}
+type Expect msg
+    = ExpectJson (Http.Expect msg)
```

Comment:
🎨 could we add a comment explaining for explicitly that "this technically can contain any expect, but we only expose public constructors that force this to always be an `Http.ExpectJson ...`